x/wasm/internal/keeper/querier.go

```diff
@@ -27,77 +28,200 @@ const (
 	QueryMethodContractStateRaw   = "raw"
 )
 
-// ContractInfoWithAddress adds the address (key) to the ContractInfo representation
-type ContractInfoWithAddress struct {
-	// embedded here, so all json items remain top level
-	*types.ContractInfo
-	Address sdk.AccAddress `json:"address"`
-}
-
 // controls error output on querier - set true when testing/debugging
 const debug = false
 
-// NewQuerier creates a new querier
-func NewQuerier(keeper Keeper) sdk.Querier {
+// NewLegacyQuerier creates a new querier
+func NewLegacyQuerier(keeper Keeper) sdk.Querier {
 	return func(ctx sdk.Context, path []string, req abci.RequestQuery) ([]byte, error) {
+		var (
+			rsp interface{}
+			err error
+		)
 		switch path[0] {
 		case QueryGetContract:
-			return queryContractInfo(ctx, path[1], req, keeper)
+			addr, err2 := sdk.AccAddressFromBech32(path[1])
+			if err2 != nil {
+				return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, err2.Error())
+			}
+			rsp, err = queryContractInfo(ctx, addr, keeper)
 		case QueryListContractByCode:
-			return queryContractListByCode(ctx, path[1], req, keeper)
+			codeID, err2 := strconv.ParseUint(path[1], 10, 64)
+			if err2 != nil {
+				return nil, err2
+			}
+			rsp, err = queryContractListByCode(ctx, codeID, keeper)
 		case QueryGetContractState:
 			if len(path) < 3 {
 				return nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "unknown data query endpoint")
 			}
-			return queryContractState(ctx, path[1], path[2], req, keeper)
+			return queryContractState(ctx, path[1], path[2], req.Data, keeper)
 		case QueryGetCode:
-			return queryCode(ctx, path[1], req, keeper)
+			codeID, err2 := strconv.ParseUint(path[1], 10, 64)
+			if err2 != nil {
+				return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "invalid codeID: %s", err2.Error())
+			}
+			rsp, err = queryCode(ctx, codeID, keeper)
 		case QueryListCode:
-			return queryCodeList(ctx, req, keeper)
+			rsp, err = queryCodeList(ctx, keeper)
 		default:
 			return nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "unknown data query endpoint")
 		}
+		if err != nil {
+			return nil, err
+		}
+		if rsp == nil || reflect.ValueOf(rsp).IsNil() {
+			return nil, nil
+		}
+		bz, err := json.MarshalIndent(rsp, "", "  ")
+		if err != nil {
+			return nil, sdkerrors.Wrap(sdkerrors.ErrJSONMarshal, err.Error())
+		}
+		return bz, nil
 	}
 }
 
-func queryContractInfo(ctx sdk.Context, bech string, req abci.RequestQuery, keeper Keeper) ([]byte, error) {
-	addr, err := sdk.AccAddressFromBech32(bech)
-	if err != nil {
-		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, err.Error())
+type grpcQuerier struct {
```

Comment:
I'd be curious for a quick explanation of the differences between the "LegacyQuerier" and the new grpc querier.

I am in the process of adding tendermint rpc queries to CosmJS and it would be great to understand this.

Reply:
`LegacyQuerier` is the handler for the the old rest queries in LCD. The grpc querier is a new endpoint to talk protobuf instead. I have not started it, yet so please consider the code WIP. There are also no tests for this.

Reply:
Please note the `wasmcli  rest-server` command is not supported in this branch anymore.